PullRequst for Issue6: Server App crashed when sending request from TCPClient

XuetaoLiu-S opened the pull request:
https://github.com/acquaman/AcquamanDataServer/issues/6

The crash is caused by the empty count field sent from the TCPClient.

Solution:
- Added a protection in the server app, if no data field, the request will be ignored
- for StartTime + Count request type, if count is not provided, default it as 1.

---

Note: 
This pull request is based on Branch Issue38, which should be reviewed first.

XuetaoLiu-S committed:
When the count is NOT providedd for StartTimePlusCount request type, default count as 1. Otherwise, the app will crash. Also add a protection for AMDSClientDataRequest, if data_.count() is 0, ignore the request

XuetaoLiu-S committed:
Merge branch 'Issue38' into Issue6

XuetaoLiu-S committed:
refine the error signal of AMDSTCPDataServer

XuetaoLiu-S committed:
- Refine the AppSetting for AMDS, save the settings to ini file;
- Refine the initialization logic

XuetaoLiu-S committed:
Set the data count in the proper position (when creating the count).

XuetaoLiu-S committed:
Merge branch 'Issue6' into Issue7

XuetaoLiu-S committed:
Code refactor

XuetaoLiu-S committed:
Print error when failed to parse type

XuetaoLiu-S committed:
add brackets for the setCount() function to avoid confusion

add brackets for the setCount() function to avoid confusion

XuetaoLiu-S committed:
- Add the RelativeCountPlusCount client request type and the logic to
handle the client request
- Add the TCPClient project

XuetaoLiu-S committed:
Merge branch 'Issue6' into Issue7

XuetaoLiu-S committed:
some more code refactors

XuetaoLiu-S committed:
code refactor: move the class registration to a single entry, so that it is easier to maintain the server/client project

XuetaoLiu-S committed:
code refactor: some code formatting changes

XuetaoLiu-S committed:
Code refactor:
remove hard-coded BufferGroup handling in AMDSCentralServer

XuetaoLiu-S committed:
move connect signal of buffergroup to the initialization codes so that the server won't connect the same signal many times

XuetaoLiu-S committed:
Added the feature for StartTimeToEndTime data request type

XuetaoLiu-S committed:
add PrintData function for AMDSClientRequest and implement it in the
AMDSClientStatisticsRequest class

XuetaoLiu-S committed:
add a general populateData function for AMDSBufferGroup, so that we can share the codes

XuetaoLiu-S committed:
Add the feature for MiddleTimePlusCountBeforeAndAfter message type

XuetaoLiu-S committed:
fix a bug for the MiddleTimePlusCountBeforeAndAfter

iainworkman commented:
Looks good to me. I would maybe suggest an addition to the documentation for start time + count request type which indicates that it the count value floors at 1, but other than that everything looks great.

dretrex commented on source/AMDSThreadedTCPDataServer.cpp:
```diff
@@ -0,0 +1,50 @@
+	connect(server_, SIGNAL(error(quint8,quint16,QString)), this, SIGNAL(error(quint8,quint16,QString)));
+
+	server_->moveToThread(thread_);
+	thread_->start(QThread::LowPriority);
```
Can you explain the idea behind the priority choice?  Is there a particular reason for a low priority vs any of the other ones?

dretrex commented on source/AMDSThreadedTCPDataServer.cpp:
```diff
@@ -0,0 +1,50 @@
+AMDSThreadedTCPDataServer::~AMDSThreadedTCPDataServer()
+{
+	if (thread_->isRunning())
+		thread_->terminate();
```
@iainworkman noticed this in a diffferent pull request, but you should probably be using `QThread::quit()` instead.

XuetaoLiu-S committed:
Update the document of the AMDSClientStartTimePlusCountDataRequest

XuetaoLiu-S committed:
Code refactor: rename printData() to validateRespose(); refine the codes in AMDSTCPDataServer

dretrex commented:
Rather than say that the default value is 1, it's more appropriate to say that the value must be _at least_ 1 because if it's 0 then it ceases to work correctly.

XuetaoLiu-S committed:
- Finished the continuous message support and update interval
- partial implemetation of timeout message and handshake message

XuetaoLiu-S committed:
Update the document for the count_ attribute. It must be >0

XuetaoLiu-S committed:
refine the connection for AMDSClientContinuousDataRequest

XuetaoLiu-S committed:
Some small code refactor

XuetaoLiu-S committed:
- reimplement the client end connection
- Add AMDSClientTCPSocket class to manage the TCPSocket connect
- Add AMDSClientTCPSocketManager class to manage the TCPSocket connect

XuetaoLiu-S committed:
Finalize the codes for supporting multiple connections on both the
server side and the client side

XuetaoLiu-S committed:
Refine the codes of TCOSocketManager

XuetaoLiu-S committed:
Remove the release connection call from AMDSClient for handshake
message, which will in turn make hankshake failed

XuetaoLiu-S committed:
Minor changes based on comments

XuetaoLiu-S committed:
When the data size is big, the server will send the message in several packages. Therefore, the client side has to adjust the logic to process the incoming data: when bytesAvailable is less than expected size, we will buffer the current data and only parse it when all data are received.

XuetaoLiu-S committed:
Fix a bug for waiting for more data issue

XuetaoLiu-S committed:
solve confliction to merge Issue7

XuetaoLiu-S committed:
Update docstrings for new added variables

XuetaoLiu-S committed:
Fix the broken build

dretrex commented:
This can be merged.  @XuetaoLiu-CLS, I'll let you do the final merge/delete branch step on these pull requests since you know the appropriate ordering.  @iainworkman and I will make our comments and give them the thumbs up and then you can merge in at your convenience.  Sound good?

dretrex commented:
Time Estimate: 1 minute

XuetaoLiu-S committed:
Refine the code to make it clearer

XuetaoLiu-S committed:
Refine the code to make it more readable

XuetaoLiu-S committed:
remove one unused comment

XuetaoLiu-S committed:
Solve the conflict with Issue7

XuetaoLiu-S committed:
change updateInterval to quint32; remove unused method isOrignalRequest for continous message

XuetaoLiu-S committed:
merge base branch in to solve conflict

XuetaoLiu-S committed:
merge base branch in to solve conflict

XuetaoLiu-S committed:
try to close the socket smoothly

XuetaoLiu-S committed:
merge base branch to solve conflict

XuetaoLiu-S committed:
Merge pull request #50 from acquaman/Issue49

Pull Request for Issue49: fix the broken continuous request when the data is too large

XuetaoLiu-S committed:
Merge pull request #51 from acquaman/Issue12

PullRequest for Issue12 and issue15: client handle multiple incoming stream and handsake/timeout feature

XuetaoLiu-S committed:
Merge pull request #56 from acquaman/Issue11

PullRequest for Issue11: finish continuous data request connection

XuetaoLiu-S committed:
Merge pull request #55 from acquaman/Issue10

  Edit PullRequest for Issue10: add update interval support for continuous message type

XuetaoLiu-S committed:
merge base branch in to solve conflict

XuetaoLiu-S committed:
Merge pull request #54 from acquaman/Issue9

PullRequst for Issue9: add message type MiddleTime plus Count Before and Afte

XuetaoLiu-S committed:
Merge pull request #57 from acquaman/Issue42

Pull Request for Issue42: solve the AMDS server multiple data process called issue

XuetaoLiu-S committed:
Merge branch 'Issue7' into Issue8

XuetaoLiu-S committed:
Merge branch 'Issue7', remote-tracking branch 'origin' into Issue8

XuetaoLiu-S committed:
Merge pull request #53 from acquaman/Issue8

PullRequest for Issue8: add feature for StartTime to EndTime request type

XuetaoLiu-S committed:
Merge pull request #52 from acquaman/Issue7

PullRequest for Issue7: implement the RelativeCountPlusCount client request type
